libcontainer/nsenter/nsexec.c

```diff
@@ -501,6 +506,12 @@ void join_namespaces(char *nslist)
 		struct namespace_t *ns = &namespaces[i];
 		int flag = nsflag(ns->type);
 
+		if (flag == CLONE_NEWUSER) {
```

Comment:
I try to use `retsnoop` to check and found this in v6.5 kernel

```
21:47:54.401485 -> 21:47:54.401526 TID/PID 69055/69055 (runc:[2:INIT]/runc:[2:INIT]):

FUNCTION CALLS                                         RESULT    DURATION
----------------------------------------------------   --------  --------
→ __x64_sys_setxattr
    → path_setxattr
        → setxattr
            → do_setxattr
                → vfs_setxattr
                    → security_inode_setxattr
                        → selinux_inode_setxattr
                            ↔ inode_owner_or_capable   [false]    1.100us
                        ← selinux_inode_setxattr       [-EPERM]   2.700us
                    ← security_inode_setxattr          [-EPERM]   8.900us
                ← vfs_setxattr                         [-EPERM]  15.300us
            ← do_setxattr                              [-EPERM]  21.400us
        ← setxattr                                     [-EPERM]  24.000us
    ← path_setxattr                                    [-EPERM]  29.500us
← __x64_sys_setxattr                                   [-EPERM]  31.700us
```

https://github.com/torvalds/linux/blob/2dde18cd1d8fac735875f2e4987f11817cc0bc2c/security/selinux/hooks.c#L3151

Is it related to idmap here?

Reply:
Maybe the net ns path created in containerd is not owned by the user ns?

Reply:
the path is about rootfs/dev/mqueue instead of net ns.
The net ns is created by new user ns.

The patch works on my local. But it's still mystery to me.

Reply:
Based on your patch, the change works on my local as well.

```diff
diff --git a/libcontainer/nsenter/nsexec.c b/libcontainer/nsenter/nsexec.c
index 74e15b96..350a17fe 100644
--- a/libcontainer/nsenter/nsexec.c
+++ b/libcontainer/nsenter/nsexec.c
@@ -505,6 +505,11 @@ void join_namespaces(char *nslist)
                if (setns(ns->fd, flag) < 0)
                        bail("failed to setns into %s namespace", ns->type);

+               if (flag == CLONE_NEWUSER) {
+                       if (setresuid(0, 0, 0) < 0)
+                               bail("failed to become root in user namespace");
+               }
+
                close(ns->fd);
        }
```

Reply:
Based on result of `inode_owner_or_capable` in linux, it looks like we use wrong fsuid to create `/dev/`, which isn't align with user mapping.

Reply:
So, the core reason is the second one in my comit msg?
2. Should also become root in the namespace, if we have joined a user ns
path like what we do in unsharing a new user ns.

Reply:
> So, the core reason is the second one in my comit msg?

Your comment is correct! Yes!

Reply:
> Based on your patch, the change works on my local as well.

But It can't work on my fedora 40.

Reply:
I am trying this in https://github.com/containerd/containerd/pull/10893 and will try locally, too.

Reply:
@lifubang I'm using Ubuntu 22'04 with 6.5 kernel. Would you mind to use retsnoop to show the stack trace ? Thanks

Reply:
Works for me locally in Fedora 40 Vagrant with SELinux enabled 

```
[root@localhost containerd]# runc --version
runc version 1.2.0+dev
commit: v1.0.0-1706-g7169f2d9
spec: 1.2.0
go: go1.23.2
libseccomp: 2.5.5
[root@localhost containerd]# sestatus 
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
SELinux root directory:         /etc/selinux
Loaded policy name:             targeted
Current mode:                   enforcing
Mode from config file:          enforcing
Policy MLS status:              enabled
Policy deny_unknown status:     allowed
Memory protection checking:     actual (secure)
Max kernel policy version:      33
[root@localhost containerd]# CONTAINERD_RUNTIME=io.containerd.runc.v2 FOCUS=TestPodUserNS ./script/test/cri-integration.sh
....
--- PASS: TestPodUserNS (5.24s)
    --- PASS: TestPodUserNS/volumes_permissions (1.32s)
    --- PASS: TestPodUserNS/fails_with_several_mappings (0.01s)
    --- PASS: TestPodUserNS/userns_uid_mapping (1.34s)
    --- PASS: TestPodUserNS/userns_gid_mapping (1.31s)
    --- PASS: TestPodUserNS/rootfs_permissions (1.25s)
PASS
```

Reply:
Oh, yes, it can work, I put the code in the wrong place.
So, we should change this PR to let the change as little as possible?

Reply:
Changed.
Thanks all of you @fuweid, @kolyshkin, and @AkihiroSuda for test.